### PR TITLE
Image drawing sf::Sprite respects transformation

### PR DESCRIPTION
**Pull request by DNKpp**
A ``sf::Sprite`` drawn via Image function doesn't respect the transformaiton. That's what I wanted to fix.
This pull request is rather WIP, but this way its easier to talk about improvements and actual behaviour.

The sprite transformation works (proof below) but I had to apply some changes to the interface. This is not a big deal, because it should be easy to simply add my function  and call it from the previous two overloads, thus no breaking change necessary. But here is my second struggle. The positioning and size behaviour of the ``sf::RenderTexture``, ``sf::Texture`` and ``sf::Sprite`` now seem to differ quite a bit. I wasn't able to determine how a ``size`` param should change the appearance of such a sprite, as a sprite itself has a complete transformation. I decided to get rid of that overload and instead add a ``sf::Transform`` param, because this is much more flexible than the provided interface of a ``sf::Sprite``. With a ``sf::Transform`` its possible to perform transformations in different order; the ``sf::Sprite`` uses a fixed ordering.

Looking forward for any feedback!

image

### CODE REVIEW

**Comment by DNKpp** [comment]
Well, I think I solved most of my struggles. I simply scaled the sprite to ``size`` after all other transformations have been applied. Additionally I improved the handling of the sprite, as it is now properly handled by ImGui (mostly copy-pasted from original ``ImGui::Image`` function).
One question is still unsolved. How to handle the border? Is it expected to apply transformations to the border rect or should it stay uneffected? Currently it is transformed, but I wonder if there ever is an actual usecase for this... (look at the picture above, its very far from prett) On the other handside this is the more difficult part. If one simply wants a border around the sprite frame, than it should be very easy to do so, without expecting the function to do it.

File: imgui-SFML.cpp [hunk 1]
```diff
@@ -688,39 +688,61 @@ void Image(const sf::RenderTexture& texture, const sf::Vector2f& size, const sf:
...
-    const sf::Texture& texture = *texturePtr;
-    ImTextureID textureID = convertGLTextureHandleToImTextureID(texture.getNativeHandle());
-    sf::Vector2f textureSize = static_cast<sf::Vector2f>(texture.getSize());
-    const sf::IntRect& textureRect = sprite.getTextureRect();
+    ImRect itemBB(window->DC.CursorPos.x, window->DC.CursorPos.y,
+        window->DC.CursorPos.x + size.x, window->DC.CursorPos.y + size.y);
+    if (borderColor.a > 0)
+    {
+        itemBB.Max.x += 2;
+        itemBB.Max.y += 2;
```
Comment (eliasdaler) [inline imgui-SFML.cpp:720]: Why is this needed?

File: imgui-SFML.cpp [hunk 2]
```diff
@@ -688,39 +688,61 @@ void Image(const sf::RenderTexture& texture, const sf::Vector2f& size, const sf:
...
+    }
+
+    ItemSize(itemBB);
+    if (!ItemAdd(itemBB, 0))
+        return;
+
+    // prepare uv coordinates
+    const sf::Vector2f textureSize = static_cast<sf::Vector2f>(sprite.getTexture()->getSize());
```
Comment (eliasdaler) [inline imgui-SFML.cpp:728]: Texture can be NULL here. Please add a check and just "return" if it's the case

File: imgui-SFML.cpp [hunk 3]
```diff
@@ -688,39 +688,61 @@ void Image(const sf::RenderTexture& texture, const sf::Vector2f& size, const sf:
...
 }
 
-void Image(const sf::Sprite& sprite, const sf::Transform& additionalTransform, const sf::Color& tintColor,
-    const sf::Color& borderColor)
+void Image(const sf::Sprite& sprite, const sf::Vector2f& size, const sf::Color& tintColor,
+    const sf::Color& borderColor) {
+	Image(sprite, size, sf::Transform::Identity, tintColor, borderColor);
+}
+
+void Image(const sf::Sprite& sprite, const sf::Vector2f& size, const sf::Transform& additionalTransform,
```
Comment (eliasdaler) [inline imgui-SFML.cpp:699]: Change "additionalTransform" to just "transform", please

File: imgui-SFML.cpp [hunk 4]
```diff
@@ -688,39 +688,61 @@ void Image(const sf::RenderTexture& texture, const sf::Vector2f& size, const sf:
...
                (textureRect.top + textureRect.height) / textureSize.y);
 
-    const ImVec2 uvs[4] =
-    {
-        uv0,
-        ImVec2(uv1.x, uv0.y),
-        uv1,
-        ImVec2(uv0.x, uv1.y)
-    };
+    sf::Transform transform = additionalTransform * sprite.getTransform();
```
Comment (eliasdaler) [inline imgui-SFML.cpp:733]: This can be named "finalTransform"

File: imgui-SFML.cpp [hunk 5]
```diff
@@ -688,39 +688,61 @@ void Image(const sf::RenderTexture& texture, const sf::Vector2f& size, const sf:
...
+	sf::FloatRect spriteRect = sprite.getLocalBounds();
+	const sf::FloatRect bounding = transform.transformRect(spriteRect);
+
+    // applies the transformations which are expected as a child of the parent window
+    sf::Transform childTransform;
+    childTransform.translate(itemBB.Min.x, itemBB.Min.y).
+        scale(size.x / bounding.width, size.y / bounding.height).
```
Comment (eliasdaler) [inline imgui-SFML.cpp:740]: Possible division by zero?

File: imgui-SFML.cpp [hunk 6]
```diff
@@ -1,5 +1,6 @@
 #include "imgui-SFML.h"
 #include <imgui.h>
+#include <imgui_internal.h>
```
Comment (eliasdaler) [inline imgui-SFML.cpp:3]: Why is this needed?

File: imgui-SFML.cpp [hunk 7]
```diff
@@ -116,6 +117,7 @@ namespace {
 ImColor toImColor(sf::Color c);
 ImVec2 getTopLeftAbsolute(const sf::FloatRect& rect);
 ImVec2 getDownRightAbsolute(const sf::FloatRect& rect);
+void toImVec2Quad(const sf::FloatRect& rect, const sf::Transform& transform, ImVec2 quad[4]);
```
Comment (eliasdaler) [inline imgui-SFML.cpp:120]: This function should return quad instead of accepting it as a parameter

File: imgui-SFML.cpp [hunk 8]
```diff
@@ -686,29 +688,85 @@ void Image(const sf::RenderTexture& texture, const sf::Vector2f& size, const sf:
...
+	Image(sprite, size, sf::Transform::Identity, tintColor, borderColor);
+}
+
+void Image(const sf::Sprite& sprite, const sf::Vector2f& size, const sf::Transform& additionalTransform,
+    const sf::Color& tintColor, const sf::Color& borderColor)
+{
+    const sf::IntRect& textureRect = sprite.getTextureRect();
+
+    // sprite without texture on empty frame cannot be drawn
+    if (!sprite.getTexture() ||
```
Comment (eliasdaler) [inline imgui-SFML.cpp:705]: This check should be done before getting textureRect

File: imgui-SFML.cpp [hunk 9]
```diff
@@ -686,29 +688,85 @@ void Image(const sf::RenderTexture& texture, const sf::Vector2f& size, const sf:
...
+        translate(-bounding.getPosition());
+    transform = childTransform * transform;
+    
+
+    if (borderColor.a > 0)
+    {
+        spriteRect.left += 1;
+        spriteRect.top += 1;
+        spriteRect.width -= 2;
+        spriteRect.height -= 2;
```
Comment (eliasdaler) [inline imgui-SFML.cpp:750]: Are the borders always 1px wide?

File: imgui-SFML.cpp [hunk 10]
```diff
@@ -686,29 +688,85 @@ void Image(const sf::RenderTexture& texture, const sf::Vector2f& size, const sf:
...
+
+    ImRect itemBB(window->DC.CursorPos.x, window->DC.CursorPos.y,
+        window->DC.CursorPos.x + size.x, window->DC.CursorPos.y + size.y);
+    if (borderColor.a > 0)
+    {
+        itemBB.Max.x += 2;
+        itemBB.Max.y += 2;
+    }
+
+    ItemSize(itemBB);
```
Comment (eliasdaler) [inline imgui-SFML.cpp:723]: Prefix all Dear ImGui functions with "ImGui" namespace please. Otherwise it's not clear that we're using ImGui's functions

**Review by eliasdaler — COMMENTED** [review]
Wow, this is pretty complicated...
I'd say that it's better to leave overload which called ImGui::Image as it's much simpler to understand and a lot more lightweight for most uses.

Yes, we'd have some duplication in "extended" version, but at least only one of the functions would break when ImGui::Image changes its algorithm. Plus, using <imgui_internal.h> makes e a bit uneasy, because it doesn't have any API compatibility promises. But I guess nothing can be done about it...

**Comment by eliasdaler** [comment]
"One question is still unsolved. How to handle the border? Is it expected to apply transformations to the border rect or should it stay uneffected? Currently it is transformed, but I wonder if there ever is an actual usecase for this... (look at the picture above, its very far from pretty) On the other handside this is the more difficult part. If one simply wants a border around the sprite frame, than it should be very easy to do so, without expecting the function to do it."

Does Dear ImGui have any widgets that can be rotated? If it does, please see how it transforms the borders.
If it doesn't - please try to create an implementation of this complex overload outside of ImGui-SFML's code. If it's possible, then maybe we should provide this as a code snippet/example. I really don't want to support a code which uses imgui_external and does so many complex things not related to the actual backend implementation, to be honest. :(

File: imgui-SFML.cpp (hunk 1 above)
Comment (DNKpp) [reply imgui-SFML.cpp:720]: Similar behaviour as in original ImGui::Image. Simply ensures, that actual drawn image is exactly the requested size, when a border shall be drawn.

File: imgui-SFML.cpp (hunk 2 above)
Comment (DNKpp) [reply imgui-SFML.cpp:728]: It is already checked in the first condition of the function...

File: imgui-SFML.cpp (hunk 8 above)
Comment (DNKpp) [reply imgui-SFML.cpp:705]: Not check necessary, as the texture rect is a value member of ``sf::Sprite``: https://github.com/SFML/SFML/blob/master/include/SFML/Graphics/Sprite.hpp

File: imgui-SFML.cpp (hunk 9 above)
Comment (DNKpp) [reply imgui-SFML.cpp:750]: Well, that's exactly what the original ``ImGui::Image`` function assumes.

**Comment by DNKpp** [comment]
> Does Dear ImGui have any widgets that can be rotated? If it does, please see how it transforms the borders. If it doesn't - please try to create an implementation of this complex overload outside of ImGui-SFML's code. If it's possible, then maybe we should provide this as a code snippet/example.
> 
Actually don't know. But the I think the best solution would be to simply draw the border independent to the transformation.
> I really don't want to support a code which uses imgui_external and does so many complex things not related to the actual backend implementation, to be honest. :(
What do you expect? Transformation handling isn't just straight forward. Creating bridge code between two independant domains, neither. Additionally you require to keep things c++03 conform.
As I have to emulate the internals of ``ImGui::Image``, inclusion of ``imgui_internal.h`` is simply necessary. Otherwise such functionas like ``ItemSize`` and ``ItemAdd`` will not be available.

File: imgui-SFML.cpp (hunk 6 above)
Comment (DNKpp) [reply imgui-SFML.cpp:3]: Due to necessary calls to ``ItemSize`` and ``ItemAdd``

File: imgui-SFML.cpp (hunk 7 above)
Comment (DNKpp) [reply imgui-SFML.cpp:120]: If you tell me how to return a c array from a function, without allocating heap mem or declaring it static, I'll do that.

**Comment by DNKpp** [comment]
I've refactored the code accordingly to your comments.
I also decided to simply draw the border without the sprite transformation. This looks like this:
image

File: imgui-SFML.cpp (hunk 5 above)
Comment (DNKpp) [reply imgui-SFML.cpp:739]: Yes, added an early return

**Comment by eigenl** [comment]
Perhaps this could be merged then? Will need this functionality personally.

**Comment by eliasdaler** [comment]
After some thought, I'm closing this.
The complexity and future maintenance cost is not worth it. 

Everything here can be done without making it a part of ImGui-SFML.
